sql/core/src/main/scala/org/apache/spark/sql/sources/commands.scala

```diff
@@ -218,24 +218,20 @@ private[sql] case class InsertIntoHadoopFsRelation(
         val partitionProj = newProjection(codegenEnabled, partitionOutput, output)
         val dataProj = newProjection(codegenEnabled, dataOutput, output)
 
-        if (needsConversion) {
-          val converter = CatalystTypeConverters.createToScalaConverter(dataSchema)
-          while (iterator.hasNext) {
-            val row = iterator.next()
-            val partitionPart = partitionProj(row)
-            val dataPart = dataProj(row)
-            val convertedDataPart = converter(dataPart).asInstanceOf[Row]
-            writerContainer.outputWriterForRow(partitionPart).write(convertedDataPart)
-          }
+        val partitionSchema = StructType.fromAttributes(partitionOutput)
+        val partConverter = CatalystTypeConverters.createToScalaConverter(partitionSchema)
```

Comment:
The Row came from Projection always have Catalyst type, so we need to convert them to Scala type. The old code is buggy, but not covered by unit tests (non-codegen Projection does not check types).

cc @liancheng